backend/crud.py

```diff
@@ -197,14 +265,231 @@ def update_schema(db: Session, schema_id: int, data: SchemaUpdateSchema) -> Sche
         db.rollback()
         raise SchemaExistsException(name=data.name, slug=data.slug)
     return sch
-        
-        
+
+
+def _get_entity_data(db: Session, entity: Entity, attr_names: List[str]) -> Dict[str, Any]:
+    data = {'id': entity.id, 'slug': entity.slug, 'deleted': entity.deleted, 'name': entity.name}
+    for attr in attr_names:
+        val_obj = entity.get(attr, db)
+        try:
+            data[attr] = val_obj.value if not isinstance(val_obj, list) else [i.value for i in val_obj]   
+        except AttributeError:
+            data[attr] = None
+    return data
+
+
+def _get_attr_values_batch(db: Session, entities: List[Entity], attrs_to_include: List[AttributeDefinition]) -> List[dict]:
+    '''Gets attr. values for list of entities by splitting attrs in
+    groups by type to select multiple attributes for all entities
+    in one query
+    '''
+    results_map = {
+        entity.id: {
+            'id': entity.id, 
+            'slug': entity.slug, 
+            'deleted': entity.deleted, 
+            'name': entity.name
+        } 
+        for entity in entities
+    }
+
+    for i in results_map.values():
+        for attr_def in attrs_to_include:
+            if attr_def.list:
+                i.update({attr_def.attribute.name: []})
+            else:
+                i.update({attr_def.attribute.name: None})
+
+    attr_groups: Dict[str, List[Attribute]] = defaultdict(list)
+    attributes = [i.attribute for i in attrs_to_include]
+    for attr in attributes:
+        attr_groups[attr.type.name].append(attr)
+    
+    ent_ids = [i.id for i in entities]
+    attr_map = {i.attribute.id: i.attribute.name for i in attrs_to_include}
+    for group, attrs in attr_groups.items():
+        value_model: Value = AttrType[group].value.model
+        q = (
+            select(value_model)
+            .where(value_model.entity_id.in_(ent_ids))
+            .where(value_model.attribute_id.in_([i.id for i in attrs]))
+        )
+        rows = db.execute(q).scalars().all()
+        for r in rows:
+            ent: dict = results_map[r.entity_id]
+            attr: str = attr_map[r.attribute_id]
+            if isinstance(ent[attr], list):
+                ent[attr].append(r.value)
+            else:
+                ent[attr] = r.value
+    
+    results = list(results_map.values())  
+    return results
+
+
+FILTER_MAP = {
+    'eq': '__eq__',
+    'lt':'__lt__',
+    'gt': '__gt__',
+    'le': '__le__',
+    'ge': '__ge__',
+    'ne': '__ne__'
+}
+
+ALLOWED_FILTERS = {
+    AttrType.STR: ['eq', 'lt', 'gt', 'le', 'ge', 'ne'],
```

Comment:
1. For matching substring (and not matching it) we have two options: `like` and `contains`. We can simulate `contains` with `like` by writing queries like `%substr%` and we have an opposite method, `not_like`, to search for strings that don't satisfy condition. On the other hand, `contains` doesn't have `not_like` alternative
2. There is a `regexp_match` method for... matching regexps and, as in case with `contains`, it doesn't have a negation variant, which means that filter inside `where` has to be negated with operator `~`. To add such negation we have to make some changes to code.

Option 1. Create `NEGATED_FILTERS` list so that when in `_query_entity_with_filters` we are iterating over filters we could check if filter is inside this list, and if it is then negate `where` condition

Option 2 sounded better inside my head, so we are left only with Option 1, but maybe you have some ideas

Reply:
1. We can expose filters `contains` and `notcontains` on the API, but use the `like` and `not_like` functions as you suggested.
2. Having the `regexp_match` might already be a good start.

> but maybe you have some ideas

I had the exact same problem with Django: It supported filtering rows by matching a RegExp, but not the opposite (filtering by not-matching). It was not too difficult to add a wrapper function in Python to make this feature available. I haven't used SqlAlchemy in a long time, but I could imagine that something similar is possible. But we don't need to worry about it now...

Reply:
> but use the like and not_like functions as you suggested.

Do you mean that when user sends `attr.contains=substr` backend should use `attr.like('%substr%')`? If so, this is not going to work (at least in current implementation) because that would require to modify query value

Reply:
Yes, that is what I had in my mind :joy: 

I'd love to see this in the MVP, but we don't need to implement it right away. We can postpone it for a while.